TensorLayout updates

lucylq opened the pull request:
Summary: Addressing comments from D67048723

Differential Revision: D68535453

dbort commented on runtime/core/tensor_layout.cpp:
```diff
@@ -0,0 +1,45 @@
+}
+} // namespace
+
+/**
+ * NOTE: the spans (sizes, dim_order) passed here must outlive the
+ * TensorLayout and all copies of it.
+ */
```
This is a message for the user creating the TensorLayout, and should go in the header.

dbort commented on runtime/core/tensor_layout.cpp:
```diff
@@ -0,0 +1,45 @@
+    const exec_aten::ScalarType& scalar_type) {
+  ssize_t n = 1;
+  for (ssize_t i = 0; i < sizes.size(); i++) {
+    ET_CHECK(sizes[i] >= 0);
```
This will crash if `sizes` contains a negative value, which we need to avoid. See my comment on the .h file about using a static factory, which would then let this helper return `Result<size_t>` to handle the error case non-fatally.

Our core tensor type suffers from the same problem, which is why the runtime has to do an [awkward pre-check](https://github.com/pytorch/executorch/blob/9bb9ad0564453bdbc656ea3502656da9eaa38c28/runtime/executor/tensor_parser_portable.cpp#L108) before creating tensors.

dbort reviewed (changes requested):
Thanks for responding to my comments on the previous PR!

lucylq commented on runtime/core/tensor_layout.h:
```diff
@@ -15,38 +15,19 @@
-class TensorLayout {
+class ET_EXPERIMENTAL TensorLayout final {
  public:
   TensorLayout(
```
thanks for explaining this!

dbort commented on runtime/core/tensor_layout.h:
```diff
@@ -10,55 +10,47 @@
-        dim_order_(dim_order),
-        scalar_type_(scalar_type),
-        nbytes_(calculate_nbytes(sizes_, scalar_type_)) {}
+  TensorLayout() = delete;
```
I don't think you need to do this since you declare a non-default ctor.

https://en.cppreference.com/w/cpp/language/default_constructor
> ### Implicitly-declared default constructor
> If there is no user-declared constructor or constructor template for a class type, the compiler will implicitly declare a default constructor as an inline public member of its class.

dbort commented on runtime/core/tensor_layout.h:
```diff
@@ -10,55 +10,47 @@
-  TensorLayout& operator=(const TensorLayout&) = default;
-  TensorLayout& operator=(TensorLayout&& other) = default;
-  ~TensorLayout() = default;
+  /** Creates a TensorLayout from the given parameters.
+   *
```
nit
```suggestion
  /**
   * Creates a TensorLayout from the given parameters.
```

dbort commented on runtime/core/tensor_layout.cpp:
```diff
@@ -0,0 +1,44 @@
+
+Result<TensorLayout> TensorLayout::create(
+    Span<const int32_t> sizes,
+    Span<const uint8_t> dim_order,
```
It'd also be a good idea to validate the dim_order entries with a loop like https://github.com/pytorch/executorch/blob/1bf20e3289da509c4344dd65d685f139fc7a172b/runtime/core/exec_aten/util/dim_order_util.h#L24-L30

And also ensure than sizes.size() == dim_order.size()

dbort commented on runtime/core/test/tensor_layout_test.cpp:
```diff
@@ -23,9 +25,12 @@ TEST(TestTensorLayout, Ctor) {
 
-  TensorLayout layout =
-      TensorLayout(ScalarType::Float, sizes_span, dim_order_span);
+  Result<TensorLayout> layout_res =
```
Now that this can fail cleanly, please add tests for the various failure cases.

dbort reviewed (approved):
Looking good! Just some final requests for new tests, but the public API looks good to me.

lucylq committed:
TensorLayout updates (#7870)

Summary:

Addressing comments from D67048723

Reviewed By: JacobSzwejbka

Differential Revision: D68535453
